### PR TITLE
[XLA] clean up generated LLVM reduction code.

### PR DESCRIPTION
**Pull request by nouiz**
This ease reading the non-optimized LLVM code.

@cheshire

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc [hunk 1]
```diff
@@ -1956,6 +1956,11 @@ static void EmitTile(
   IrArray::Index source_idx = tile_origin_index.AddOffsetToDim(
       start_offset_x, KernelMappingScheme::DimX, b);
 
+  // True when all threads will always execute all instructions.
+  // So we do not need to emit condition.
+  bool always_full_tile =
+      mapping_scheme.GetDimsInElems()[2] % tile_size_x == 0;
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: Note that tile_size_x is per thread block, wouldn't you need to divide by num_threads_x first?

File: tensorflow/compiler/xla/service/llvm_ir/kernel_support_library.cc [hunk 2]
```diff
@@ -60,11 +60,14 @@ Status KernelSupportLibrary::IfWithStatus(
     absl::string_view name, llvm::Value* condition,
     const std::function<Status()>& true_block_generator,
     const std::function<Status()>& false_block_generator) {
-  llvm_ir::LlvmIfData if_data = llvm_ir::EmitIfThenElse(condition, name, b_);
+  llvm_ir::LlvmIfData if_data = llvm_ir::EmitIfThenElse(
+      condition, name, b_, false_block_generator != nullptr);
```
Comment (cheshire) [inline tensorflow/compiler/xla/service/llvm_ir/kernel_support_library.cc]: Insert a comment on what parameter are you populating, e.g. `/*param_name=*/...`

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc (hunk 1 above)
Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: Lets dive in the detail to make sure what should be done. I could have misunderstood something.

Suppose we have a matrix of M,N and we reduce rows, so the output is of size M.

We have each thread handle up to U number. (The "unroll" factor)
We have num_threads_x the number of threads per block.
We have each block reduce a total of tile_size_x number.
To my understanding, the potential number of load from all the threads in a block : tile_size_x=num_threads_x*U.
Currently, when N isn't a multiple of tile_size_x, we launch an block that would reduce fewer elements then tile_size_x. then some threads would read invalid data. So there is a condition to check if the next element to read is within bound.

Did I miss something? Mostly, if all threads from all blocks will always full unroll, we do not need the conditions.
The [pseudo-code](https://github.com/tensorflow/tensorflow/blob/eacb0742a9816ecfcd6e2710e0d1244fdd59f654/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc#L1920) make me understand that one block can read up to tile_size_x elements.

For cols reduction, always_full_tile see to be frequently, but not always true in tests. I think the logic works too for columns reduction. All reduction tests that I found pass.

File: tensorflow/compiler/xla/service/llvm_ir/kernel_support_library.cc (hunk 2 above)
Comment (nouiz) [reply tensorflow/compiler/xla/service/llvm_ir/kernel_support_library.cc]: done.

File: tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc (hunk 1 above)
Comment (cheshire) [reply tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: Suppose we get

```
num_threads_x=32
tile_size_x=32
dims[2]=512
```

512 is divisible by 32, but not all threads will fit, as e.g. thread 15 will try to make accesses 480, 512, 544, 576, etc, and only the first two will be in bounds.

Comment (nouiz) [reply tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc]: In this case, each thread will only do 1 element. And many block will be submited.
From this line: https://github.com/tensorflow/tensorflow/blob/eacb0742a9816ecfcd6e2710e0d1244fdd59f654/tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc#L1947

We see that the number of element one thread will load is "tile_size_x / num_threads_x". So 1 in this case.

**Review by cheshire — APPROVED** [review]

**Comment by gbaned** [comment]
@nouiz Can you please resolve conflicts? Thanks!

**Comment by nouiz** [comment]
@gbaned It is now merged. So all should be good I think. Thanks for watching what is going on.

**Comment by cheshire** [comment]
> It is now merged
@nouiz No it's not, check out the "merge" message: `ArmageddonKnight pushed a commit to UofT-EcoSystem/tensorflow that referenced this pull request`

**Comment by cheshire** [comment]
@nouiz Actually I was wrong, sorry again. Something went unexpected here, let me double check what happened.

**Pull request merged**